---
title: "Assignment #2 - Gene Expression Analysis"
author: "Raphaëlle Luisier"
date: "April 2025"
output:
  html_document:
    theme: paper
#paper readable
    highlight: monochrome
    code_folding: hide
    toc: true
    toc_depth: 2
#    toc_float: true
    number_sections: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache =TRUE)
knitr::opts_chunk$set(fig.width=16, fig.height=10, figure.align = "center") 
rm(list = ls())


if (!require("BiocManager", quietly = TRUE))
  BiocManager::install("BiocManager")
if (!require("DESeq2", quietly = TRUE))
  BiocManager::install("DESeq2")
library("DESeq2")
if (!require("ape", quietly = TRUE))
  BiocManager::install("ape")
library("ape")
if (!require("limma", quietly = TRUE))
  BiocManager::install("limma")
library("limma")
if (!require("geneplotter", quietly = TRUE))
  BiocManager::install("geneplotter")
library("geneplotter")
if (!require("gprofiler2", quietly = TRUE))
  BiocManager::install("gprofiler2")
library("gprofiler2")
if (!require("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")
library("edgeR")
if (!require("gplots")) {
   install.packages("gplots", dependencies = TRUE)
library(gplots)
}

```


# Biological questions and data-set {.tabset}

## Biological Question

Amyotrophic Lateral Sclerosis (ALS) is a fatal, currently incurable neurodegenerative disorder, typically manifesting between the ages of 40 and 60. It is characterized by the progressive degeneration of motor neurons, leading to paralysis and, ultimately, death within 3–5 years of disease onset. The underlying causes of ALS remain largely unknown, and no effective treatments are currently available.

A molecular hallmark of ALS is the mislocalization of RNA-binding proteins to the cytoplasm, where they form pathological aggregates. These proteins, which normally function in the nucleus as key regulators of mRNA metabolism, contribute to disease pathology through both a nuclear loss of function and a toxic cytoplasmic gain of function.

**In this assignment, your goal is to investigate whether ALS-related mutations lead to changes in gene expression during the differentiation of induced pluripotent stem cells (iPSCs) into motor neurons.**


## Data-set

The data required for this assignment are similar to those used in Lecture 8 & 9 and can be downloaded from [Zenodo](https://zenodo.org/records/10944047). The high-throughput poly(A) RNA-seq data are described in [Neeves et al., *Brain* (2022)](https://pubmed.ncbi.nlm.nih.gov/33693641/).

These data are derived from **cytoplasmic** fractions of human induced pluripotent stem cells (hiPSCs; day 0), neural precursors (NPCs; day 3 and day 7), ‘patterned’ precursor motor neurons (ventral spinal cord; pMN; day 14), post-mitotic but electrophysiologically immature motor neurons (MN; day 22), and electrophysiologically active motor neurons (mMNs; day 35). Samples were generated from both **healthy** individuals and **ALS** patients.

![Schematic depicting the iPSC differentiation strategy for motor neurogenesis. Arrows indicate sampling time-points in days when cells were fractionated into nuclear and cytoplasmic compartments prior to deep (polyA) RNA-sequencing. Four iPSC clones were obtained from four different healthy controls and three iPSC clones from two ALS patients with VCP mutations: R155C and R191Q; hereafter termed VCPmu. Induced-pluripotent stem cells (iPSC); neural precursors (NPC); “patterned” precursor motor neurons (ventral spinal cord; pMN); post-mitotic but electrophysiologically inactive motor neurons (MN); electrophysiologically active MNs (mMN).](./protocol.png)

The gene expression count data were obtained using [Kallisto](https://pachterlab.github.io/kallisto/about) (Bray et al., 2016), aligned against the Gencode hg38 release of the *Homo sapiens* transcriptome.


```{r load_data,warning=FALSE,fig.width=8, fig.height=4}
load("./data_09_04_2024.RData")
#Data: 
# myE_ge                  : raw gene expression count matrix 
# info                    : sample annotation (data-frame)
# ttg                     : rows (genes) annotation

# Focus on CTRL samples for this session
sel_samples <- which(info$Fraction=="Cytoplasmic")
myE_ge      <- myE_ge[,sel_samples]
info        <- info[sel_samples,]
info$group  <- factor(paste(info$mutant,info$DIV,sep="_"),levels=unique(paste(info$mutant,info$DIV,sep="_")))


#Make some nice colors to facilitate the visualisation of time-points
#Colors per time-points
mytime                 <- factor(as.character(info$DIV),levels=c(0,3,7,14,22,35))
mycols_days            <- c("#CCFF00","#33CC33","#669999","#6699FF","#3300FF","#CC33CC")
names(mycols_days)     <- c(0,3,7,14,22,35)
mycols                 <- unlist(lapply(info$DIV,function(Z)return(mycols_days[match(as.character(Z),names(mycols_days))])))
#Colors per patient type (control versus ALS==VCP mutant)
mycond                 <- info$mutant
mycols_cond            <- c("#6699FF","#CC33CC")
names(mycols_cond)     <- c("CTRL","VCP")
mycols1                <- unlist(lapply(info$mutant,function(Z)return(mycols_cond[match(as.character(Z),names(mycols_cond))])))

#Log-transform your count data
myE_gel <- log2(1+myE_ge)
```


# Task 1: Identification of reliably expressed genes (2/10 pts)

Your first task is to identify the **reliably expressed genes per condition**.  
> ⚠️ *Note: Conditions in this assignment differ from those used in previous exercises — make sure to define them correctly (e.g., time points, genotypes, or their combinations).*

## Steps:
1. For each condition, identify reliably expressed genes. You may define "reliable expression" based on a minimum count threshold (e.g., counts > 10 in most replicates).
2. Visualize the number of reliably expressed genes **per condition** in a plot.
3. Report the **total number of reliably expressed genes** across all conditions (e.g., union or intersection depending on your criteria).
4. Please discuss whether there are differences between control and ALS patients in the number of reliably expressed genes, and whether this result was expected — using a maximum of two sentences.


```{r selection_threshold,eval=TRUE,fig.width=10,fig.height=3}
layout(matrix(c(1,1,2,3),ncol=4,nrow=1,byrow=TRUE))
# 1 Fit bimodal distribution
#bimdens              <- ...

# 2. Identify limit that discriminate foreground from background for each samples
#Lims                 <- ...

# 3. Plot the distribution of the pre-filtered gene counts along side the limits

# 4. Select reliably expressed genes in each sample

# 5. Select reliably expressed genes in each group

# 6. Select reliably expressed genes to be considered in this study
#is_expressed_global  <- ...

# 7. Create data-table with reliably expressed genes
myE_gelf             <- myE_gel[is_expressed_global,]


temp <- matrix(no_reliably_expressed_genes_group,nrow=2,byrow=TRUE)
rownames(temp)<- c("CTRL","ALS")
boxplot(no_reliably_expressed_genes_samples,las=1,ylab="# reliably expressed per sample")
barplot(t(temp),beside=TRUE,las=1,ylab="# reliably expressed genes per group",col=mycols_days)
legend("top",ncol=1,leg=c(0,3,7,14,22,35),col=mycols_days,bty="n",cex=0.8,pch=15)
```


***Answer***


## Quantile normalisation
```{r normalisation,eval=TRUE,fig.width=7,fig.height=4}
#Quantile normalisation
myE_gen               <- limma::normalizeQuantiles(myE_gelf)
par(mfrow=c(2,2),mar=c(4,4,2,2))
geneplotter::multidensity(myE_gelf,leg=FALSE,main="Prior normalisation",las=1,xlab="read count [log2]")
geneplotter::multidensity(myE_gen,leg=FALSE,main="Quantile normalised",las=1,xlab="read count [log2]")


boxplot(myE_gelf,outline=FALSE,las=1,ylab="read count")
boxplot(myE_gen,outline=FALSE,las=1)

```



# Task 2: Unsupervised Hierarchical Clustering Analysis (2/10 pts)

Perform hierarchical clustering analysis of the samples.  
*Hint: Use the different color palettes generated above to annotate your heatmap or dendrogram.*

Discuss the clustering in terms of the covariates, specifically, **differentiation stage** versus **ALS mutation status**,  using a maximum of two sentences.


```{r hc_all_dist_corr_mannathan,fig.width=5,fig.height=5}
CEX=0.7
par(mfrow=c(1,2))

```


**Answer**  



# Singular Value Decomposition (SVD)

## Task 3: How Many Principal Components to Be Considered (2/10 pts)

You will now perform a Singular Value Decomposition (SVD) analysis on the gene expression data and determine how many principal components are required to capture **90% of the total variance** in gene expression.

```{r investigate_SVD,warning=FALSE}
getFractionVariance<- function(mySVD){
  return(mySVD$d*mySVD$d/sum(mySVD$d*mySVD$d))
}

getShannonEntropy <- function(pip)return(-1*sum(pip*log10(pip))/log10(length(pip)))

ScrePlot <- function(pi,dp){
  barplot(pi,las=1,cex.main=0.7,cex.axis=1.0,col="black")
  mtext(side = 1, line = 2, text ="principal components", col = "black",cex=0.7, font=1)
  mtext(side = 2, line = 2, text ="Fraction of explained variance", col = "black",cex=0.7, font=1)
  mtext(side = 3, line = -2, text = paste("Shannon Entropy = ",round(dp,digits=3)), col = "black",cex=0.7, font=1)
}
```

```{r calculate_SVD,warning=FALSE,fig.width=7, fig.height=3}
# 1. SVD on the Gene Expression


#2. Calculate SE and fraction of variance captured


#2. Remove the first principal component:

#3. Calculate again the SVD

#4. Calculate SE and fraction of variance captured

#5. Scre plots before and after removal of the first component 


#6. How many principal components to capture 90% of the variance

#print(paste("The first",..., "caputre 90% of the variance in gene expression"))
```

## Task 4: PCA versus UMAP dimensionality reduction

### Task 4.1: Generate PCA Plots of the First 4 Components (1/10 pts)

Generate PCA plots for the first four principal components of the gene expression data.  
Make sure to **label the axes** (e.g., "PC1 (variance%)") and use **appropriate colors** to reflect relevant covariates (e.g., time point, genotype, or condition).


```{r pca_svd_col_time,warning=FALSE,fig.width=10, fig.height=5}
par(mfrow=c(4,4),mar=c(4,4,1,1))
#...
```

### Task 4.2: Contrast with UMAP Plot (0.5/10 pts)

Discuss the effects of the **number of neighbors** and **minimum distance** parameters on the UMAP plot in one or two sentences.

```{r UMPA_coeffs,warning=FALSE,fig.width=10, fig.height=5}
library(umap)
par(mfrow=c(3,4),mar=c(3,3,2,2))
for(n_nb in c(4,12,26)){
  for(mind in c(0.01, 0.05,0.1,0.9)){
    plot(umap(t(myE_gen),n_neighbors = n_nb,min_dist = mind)$layout,pch=19,cex=0.8,col=mycols,xlab="",ylab="",xaxt="n",yaxt="n",frame=FALSE)
    mtext(side=3,line=0,text=paste("#neighbors=",n_nb),cex=0.8)
    mtext(side=3,line=1,text=paste("min distance=",mind),cex=0.8)
  } 
}
```

***Answer***

### Task 4.3: Compare with UMAP Plot (0.5/10 pts)

Discuss the differences between **SVD** (or PCA) and **UMAP** in one or two sentences.

```{r UMPA_PCA,warning=FALSE,fig.width=10, fig.height=2}
par(mfrow=c(1,4),mar=c(4,4,2,2))
plot(SVD_2$v[,1], SVD_2$v[,2],col=mycols,xlab="v1",ylab="v2",pch=19,cex=1.5,las=1,frame=FALSE)
legend("topright",pch=19,col=mycols_days,ncol=1,leg=paste("d",names(mycols_days)),cex=0.6)

for(n_nb in c(4,12,26)){
mind=0.9
plot(umap(t(myE_gen),n_neighbors = n_nb,min_dist = mind)$layout,pch=19,cex=0.8,col=mycols,xlab="",ylab="",xaxt="n",yaxt="n",frame=FALSE)
mtext(side=3,line=0,text=paste("#neighbors=",n_nb),cex=0.8)
mtext(side=3,line=1,text=paste("min distance=",mind),cex=0.8)
}
```

***Answer***

## Task 5: ALS Effect During Development (2/10 pts)

### Task 5.1: Components Over Time (2/10 pts)

Plot the principal components over time using the first left singular vectors.  
Be sure to **separate control versus ALS samples** in your visualization.  
Discuss in **one sentence** whether an ALS-related effect is detected.


```{r component_over_time,warning=FALSE,fig.width=10, fig.height=4}
#1. Select the loading of the control samples as well as info; create a new grouping
#...
#2. Select the loading of the ALS/VCP samples as well as info; create a new grouping
#...

#3. Compute the average and SD per group per singular vectors
#...

#4. Plot component over time
error.bar <- function(x, y, upper, lower=upper, length=0.1,...){
  if(length(x) != length(y) | length(y) !=length(lower) | length(lower) != length(upper))
    stop("vectors must be same length")
  arrows(x,y+upper, x, y-lower, angle=90, code=3, length=length, ...)
}
days                <- c(0,3,7,14,21,35)
CEX<- 0.7
par(mfrow=c(1,4),mar=c(5,5,2,0),oma=c(2,2,2,2))

for(i in c(1:4)){
    MIN=min(0,min(c(myMean_ct[i,]-mySD_ct[i,],myMean_ct[i,]-mySD_ct[i,])))
    MAX=max(c(myMean_ct[i,]+mySD_ct[i,],myMean_ct[i,]+mySD_ct[i,]))
    
    plot(days,myMean_ct[i,],pch=19,type="b",lty=2,ylim=c(MIN,MAX),las=1,frame="F",xlab="time [days]",cex=1.0,cex.axis=CEX,cex.lab=CEX,ylab="")
    mtext(side=2,line=3,text=paste("PC",i,sep=""),cex=CEX)
    mtext(side=3,line=0,text=paste(round(100*getFractionVariance(SVD_2)[i],digit=2),"% of variance"),cex=CEX)
    grid()
    error.bar(x=days, y=myMean_ct[i,], upper=mySD_ct[i,],  length=0.05)
    abline(h=0,col="red",lty=2)
    
    points(days,myMean_mut[i,],pch=19,type="b",cex=1.0,cex.axis=CEX,col="grey")
    lines(days,myMean_mut[i,],lty=2,col="grey")
    error.bar(x=days, y=myMean_mut[i,], col="grey",upper=mySD_mut[i,],  length=0.05)
    
    if(i==1){
      legend("topright",pch=19,col=c("black","grey"),cex=0.6,leg=c("CTRL","VCP"),ncol=2)
    }
}

```

**Answer**  


### Task 5.2 – Bonus: Investigate the Biological Meaning of the Component of Interest (2/10 pts)

Having identified a component associated with ALS, you are now invited to further investigate the **genes most strongly contributing to this component** and explore their associated **biological pathways**.  
Please provide:

- A description of how the genes were selected (e.g., based on loading values).
- The results of a pathway enrichment or gene ontology analysis.


#### Bonus 1 – Extraction of Top-Contributing and Least-Contributing Genes (1 pt)

Extract the genes that contribute most strongly (both positively and negatively) to the component of interest, as well as those that contribute the least.  
These genes can be identified based on their **loading values** (right singular vectors) in the SVD analysis.


```{r extract_most_contributing,warning=FALSE,fig.width=7, fig.height=7,eval=TRUE}

```

#### Bonus 2 – Visual Inspection of Gene Behaviors (1 pt)

Please provide a visualization of your choice to illustrate the behavior of the genes most strongly associated with the component of interest.  
*Remember:* There are both positively and negatively contributing genes, so consider showing both directions of contribution (e.g., using a heatmap, expression trends, or correlation plots).




